### PR TITLE
Revert changes of #1478 and #1456

### PR DESCRIPTION
**Pull request by Tortar**
Reverts projectmesa/mesa#1478 and parts of projectmesa/mesa#1456 because they introduce the bug discussed in #1515

### CODE REVIEW

**Comment by Corvince** [comment]
Thanks for the work. Would be awesome if we would have a test case for this, esp. the removing-while-running part. Do you think you could add those as well? No worries if you don't have time for that, we don't have them now so this is not a requirement for this PR to be merged

**Comment by Tortar** [comment]
Will do that for sure, but right now I can't unfortunately

File: mesa/time.py [hunk 1]
```diff
@@ -185,18 +186,16 @@ def __init__(
...
-        agent_keys = self._agents.keys()
+        agent_keys = list(self._agents.keys())
         if self.shuffle:
-            agent_keys = list(agent_keys)
             self.model.random.shuffle(agent_keys)
         for stage in self.stage_list:
             for agent_key in agent_keys:
                 getattr(self._agents[agent_key], stage)()  # Run stage
             # We recompute the keys because some agents might have been removed
             # in the previous loop.
```
Comment (rht) [inline mesa/time.py:196]: Needs to document why the casting to `list` is necessary.

Comment (Tortar) [reply mesa/time.py:202]: done

**Comment by rht** [comment]
LGTM

**Pull request merged**